www/caddy/src/opnsense/mvc/app/models/Pischem/Caddy/Caddy.php

```diff
@@ -0,0 +1,177 @@
+            } else {
+                $defaultPorts = [$fromPort];
+            }
+
+            foreach ($defaultPorts as $port) {
+                // Create a unique key for domain/subdomain-port combination
+                $comboKey = $fromDomainOrSubdomain . ':' . $port;
+
+                // Check for duplicate combinations
+                if (isset($combos[$comboKey])) {
+                    // Use dynamic $key for message referencing
+                    $messages->appendMessage(new Message(
+                        "Duplicate entry: The combination of $type '$fromDomainOrSubdomain' and port '$port' is already used. Each $type and port pairing must be unique.",
+                        $type === 'domain' ? $key . ".FromDomain" : $key . ".FromDomain", // Adjusted to use dynamic key
+                        "Duplicate" . ucfirst($type) . "Port"
+                    ));
+                } else {
+                    $combos[$comboKey] = true;
+                }
+            }
+        }
+    }
+
+    // 3. Check that subdomains are under a wildcard or exact domain
+    private function checkSubdomainsAgainstDomains($subdomains, $domains, $messages)
+    {
+        $wildcardDomainList = [];
+        foreach ($domains as $domain) {
+            if ((string) $domain->enabled === '1') {
+                $domainName = (string) $domain->FromDomain;
+                if (str_starts_with($domainName, '*.')) {
+                    $wildcardBase = substr($domainName, 2);
+                    $wildcardDomainList[$wildcardBase] = $domainName;
+                }
+            }
+        }
+
+        foreach ($subdomains as $subdomain) {
+            if ((string) $subdomain->enabled === '1') {
+                $subdomainName = (string) $subdomain->FromDomain;
+                $isValid = false;
+                foreach ($wildcardDomainList as $baseDomain => $wildcardDomain) {
+                    if ($this->endsWith($subdomainName, $baseDomain)) {
+                        $isValid = true;
+                        break;
+                    }
+                }
+
+                if (!$isValid) {
+                    $key = $subdomain->__reference; // Dynamic key based on subdomain reference
+                    $messages->appendMessage(new Message(
+                        "Invalid subdomain configuration: '$subdomainName' does not fall under any configured wildcard domain.",
+                        $key . ".FromDomain", // Use dynamic key for message referencing
+                        "InvalidSubdomain"
+                    ));
+                }
+            }
+        }
+    }
+
+    /**
+     * Checks if a string (haystack) ends with another string (needle).
+     *
+     * @param string $haystack The string to search in.
+     * @param string $needle The substring to search for at the end of $haystack.
+     * @return bool Returns true if $haystack ends with $needle, false otherwise.
+     */
+    private function endsWith($haystack, $needle)
```

Comment:
you are aware there's `str_ends_with` nowadays ? ( https://www.php.net/manual/en/function.str-ends-with.php )

Reply:
Can I change this later? I would like to keep it as it is right now. I'll put it on my list.

Reply:
This was just a little change, thanks for the hint: 532d2f4